src/high-level/matrix.lisp

```diff
@@ -0,0 +1,560 @@
+            ((subtypep type 'double-float) (cffi:mem-aptr base :double idx))
+            ((subtypep type '(complex single-float)) (cffi:mem-aptr base :float (* 2 idx)))
+            ((subtypep type '(complex double-float)) (cffi:mem-aptr base :double (* 2 idx)))
+            (t (error "Incompatible element type ~a." type))))))))
+
+(defgeneric row (matrix index)
+  (:documentation "Get row vector from a matrix")
+  (:method ((m matrix) index)
+    (check-type index alexandria:non-negative-fixnum)
+    (slice m
+           (list index 0)
+           (list (1+ index) (ncols m)))))
+
+(defgeneric column (matrix index)
+  (:documentation "Get column vector from a matrix")
+  (:method ((m matrix) index)
+    (slice m
+           (list 0 index)
+           (list (nrows m) (1+ index)))))
+
+(defgeneric mult (a b &key target alpha beta transa transb)
+  (:documentation "Multiply matrix a by matrix b, storing in target or creating a new matrix if target is not specified.
+Target cannot be the same as a or b."))
+
+(defgeneric @ (matrix &rest matrices)
+    (:documentation "Multiplication of matrices")
+  (:method (matrix &rest matrices)
+    (reduce #'mult matrices
+            :initial-value matrix)))
+
+;;; Generic matrix methods
+
+(defgeneric direct-sum (a b)
+  (:method ((a matrix) (b matrix))
+    "Compute the direct sum of A and B"
+    (let* ((arows (nrows a))
+           (acols (ncols a))
+           (brows (nrows b))
+           (bcols (ncols b))
+           (rrows (+ arows brows))
+           (rcols (+ acols bcols))
+           (result (magicl:empty (list rrows rcols)
+                                 :type '(complex double-float))))
+      (loop :for r :below arows :do
+        (loop :for c :below acols :do
+          (setf (tref result r c) (tref a r c))))
+      (loop :for r :from arows :below rrows :do
+        (loop :for c :from acols :below rcols :do
+          (setf (tref result r c) (tref b (- r arows) (- c acols)))))
+      result)))
+
+(defgeneric kron (a b &rest rest)
+  (:documentation "Compute the kronecker product of A and B")
+  (:method (a b &rest rest)
+    (let ((ma (nrows a))
+          (mb (nrows b))
+          (na (ncols a))
+          (nb (ncols b)))
+      (flet ((calc-i-j (i j) (* (tref a (floor i mb) (floor j nb))
+                                (tref b (mod i mb) (mod j nb)))))
+        (reduce #'kron rest :initial-value (into! #'calc-i-j
+                                                  (empty (list (* ma mb) (* na nb))
+                                                         :type '(complex double-float))))))))
+
+(defgeneric transpose! (matrix &key fast)
+  (:documentation "Transpose MATRIX, replacing the elements of MATRIX, optionally performing a faster change of layout if FAST is specified")
+  (:method ((matrix matrix) &key fast)
+    "Transpose a matrix by copying values.
+If fast is t then just change layout. Fast can cause problems when you want to multiply specifying transpose."
```

Comment:
```suggestion
If FAST is t then just change layout. Fast can cause problems when you want to multiply specifying transpose."
```
Also I have no idea what the second sentence means.

Reply:
I will reword that. It mostly has to do with the fact that a `FAST` transpose for matrices just changes the `shape` and the `layout`, not modifying the contents. The main issue with this is that when you do a conjugate transpose in `mult` the matrix cannot be `ROW-MAJOR` (because it would require a slow transpose prior to calling lapack)